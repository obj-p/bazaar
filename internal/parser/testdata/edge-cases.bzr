package edgecases

enum EmptyEnum {}

component EmptyComponent {}

data EmptyData {}

template EmptyTemplate {}

template EdgeCases {
    // Empty containers
    var emptyArray = []
    var emptyMap = {:}

    // Complex trailing commas
    call(
        first,
        second,
    )

    var array = [1, 2, 3,]

    // Map literals with content (lookahead=4 fixes both identifier and string keys!)
    // var map = {a: 1, b: 2,}
    // var map2 = {"a": 1, "b": 2}

    // Single annotation
    @State var count = 0

    // TODO: Multiple annotations on single construct - grammar only supports one
    // @State @Modifier(padding = 8) var count = 0
    // TODO: Revisit allowing multiple annotations on single statements

    // Complex lambda parameters with types
    call { (x int, y string) -> bool in
        return x > 0
    }

    call { () -> string in
        return "result"
    }

    call { (a, b, c) in
        // inferred types
        return a
    }

    // Complex optional chaining
    var result = obj?.method()?.property
    var indexed = array?[0]?.field
    var called = func?(param)?.result
}
