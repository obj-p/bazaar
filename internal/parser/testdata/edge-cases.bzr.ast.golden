&parser.Bazaar{
  Package: &parser.PackageDecl{
    Domain: []string{
      "edgecases",
    },
  },
  Decls: []*parser.TopLevelDecl{
    {
      Enum: &parser.EnumDecl{
        Name: "EmptyEnum",
      },
    },
    {
      Component: &parser.ComponentDecl{
        Name: "EmptyComponent",
      },
    },
    {
      Data: &parser.DataDecl{
        Name: "EmptyData",
      },
    },
    {
      Template: &parser.TemplateDecl{
        Name: "EmptyTemplate",
      },
    },
    {
      Function: &parser.FunctionDecl{
        Name: "EdgeCases",
        Block: []*parser.Stmt{
          {
            Var: &parser.VarDeclStmt{
              Dest: &parser.VariableExpr{
                Name: &"component",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Literal: &parser.Literal{
                      String: &parser.String{
                        Fragments: []*parser.StringFragment{
                          {
                            Text: &"foo",
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            Var: &parser.VarDeclStmt{
              Dest: &parser.VariableExpr{
                Name: &"data",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Literal: &parser.Literal{
                      String: &parser.String{
                        Fragments: []*parser.StringFragment{
                          {
                            Text: &"bar",
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            Var: &parser.VarDeclStmt{
              Dest: &parser.VariableExpr{
                Name: &"modifier",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Literal: &parser.Literal{
                      String: &parser.String{
                        Fragments: []*parser.StringFragment{
                          {
                            Text: &"baz",
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            Var: &parser.VarDeclStmt{
              Dest: &parser.VariableExpr{
                Name: &"package",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Literal: &parser.Literal{
                      String: &parser.String{
                        Fragments: []*parser.StringFragment{
                          {
                            Text: &"qux",
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          {
            Var: &parser.VarDeclStmt{
              Annotation: &parser.AnnotationExpr{
                Name: &"State",
              },
              Dest: &parser.VariableExpr{
                Name: &"count",
              },
              Source: &parser.Expr{
                Unary: &parser.UnaryExpr{
                  Primary: &parser.PrimaryExpr{
                    Literal: &parser.Literal{
                      Number: &"0",
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
}
